### PR TITLE
fix Issue 20990 - Optimizer should move cold branches to the end of t…

### PR DESCRIPTION
**Pull request by WalterBright**
…he function

### CODE REVIEW

File: src/dmd/backend/blockopt.d [hunk 1]
```diff
@@ -2364,4 +2365,82 @@ private void blassertsplit()
...
+        /* Not sure of effect of jumping out of a try block
+         */
+        if (b.Btry)
+            continue;
+
+        if (b.BC == BCexit)
+            continue;
+
+        if (b.Belem && !el_returns(b.Belem))
+
```
Comment (thewilsonator) [inline src/dmd/backend/blockopt.d:2387]: this can be an early continue, also extra line.

**Review by thewilsonator — COMMENTED** [review]
The test doesn't actually test anything

**Comment by WalterBright** [comment]
> The test doesn't actually test anything

What it does is tickle the optimization, so if it crashes the compiler it'll show up.

**Comment by Geod24** [comment]
> What it does is tickle the optimization, so if it crashes the compiler it'll show up.

But then, I'm sure we already have plenty of code in the testsuite that already does this.
Any test that would pass before the patch doesn't make sense to me.

**Comment by WalterBright** [comment]
> I'm sure we already have plenty of code in the testsuite that already does this.

Maybe we do, maybe we don't. I like having a specific test for it.

> Any test that would pass before the patch doesn't make sense to me.

There are many optimizer tests like that in the test suite.

**Comment by WalterBright** [comment]
The failing Win64 autotest seems to be related to https://github.com/dlang/druntime/pull/2436

**Comment by WalterBright** [comment]
Which @rainers has just fixed!

**Comment by ibuclaw** [comment]
> There are many optimizer tests like that in the test suite.

There has been some discussions of adding in a directive to scan the resultant assembler to ensure it produces what we expect it to, but there's not really any strong desire to add it.

**Review by ibuclaw — APPROVED** [review]
To my dismay, there's not even the word "cold" anywhere in the patch, so can't pretend I know what is going on here.

Checked with the following:
```
class Foo
{
    int *ptr;
    int inc;
}

class Bar
{
    Foo foo;
    int inc;
}

class Baz
{
    Bar bar;
    int inc;
}

// foo() and bar() should produce the same code when
// optimized.

void foo(Baz baz)
{
    if (baz is null)
        assert(false);
    baz.inc++;
    if (baz.bar is null)
        assert(false);
    baz.bar.inc++;
    if (baz.bar.foo is null)
        assert(false);
    baz.bar.foo.inc++;
    if (baz.bar.foo.ptr is null)
        assert(false);
    *baz.bar.foo.ptr = 42;
}

void bar(Baz baz)
{
    assert(baz);
    baz.inc++;
    assert(baz.bar);
    baz.bar.inc++;
    assert(baz.bar.foo);
    baz.bar.foo.inc++;
    assert(baz.bar.foo.ptr);
    *baz.bar.foo.ptr = 42;
}
```
And the output is what I expect it to be, so seems reasonable to me.

**Comment by ibuclaw** [comment]
However it doesn't cater for early returns instead of asserts.
```
int foo(Baz baz)
{
    if (baz is null)
        return 0x10;
    baz.inc++;
    if (baz.bar is null)
        return 0x20;
    baz.bar.inc++;
    if (baz.bar.foo is null)
        return 0x40;
    baz.bar.foo.inc++;
    if (baz.bar.foo.ptr is null)
        return 0x80;
    *baz.bar.foo.ptr = 42;
    return 0x100;
}
```
Though the above may be more reliant on branch prediction, or understanding the interdependencies between branches in order to determine what is likely or unlikely to happen.

**Comment by WalterBright** [comment]
> However it doesn't cater for early returns instead of asserts.

The optimizer is built on the presumption that the hot branch of `if-then-else` is the `then` branch.

**Pull request merged**